### PR TITLE
Fix reference data for WCS to Pixels to WCS

### LINKED ISSUE #2724: BUG: WCS to Pixels to WCS gives wrong reference data
This is a follow-up of:

* #2179 

And discovered in:

* #2712

Workflow to reproduce:

1. Use the `ImvizDitheredExample.ipynb` example notebook. Load both data.
2. Link by WCS on the Orientation plugin.
3. Click the "N-up E-left" icon so now the images in first viewer is aligned by N-up E-left.
4. Create a second viewer and load the first image only.
5. With second viewer (`imviz-1`) selected in Orientation, click on the "N-up E-right" icon so now the second viewer image is aligned by N-up E-right.
6. Now select "Pixels" in Orientation. Both viewers will go back to pixel linking and no longer appear rotated.
7. Now click "WCS" again in Orientation. See that both images do not go back to being aligned by one of the 3 possible WCS layers (see screenshot below). Despite that, when you blink, the stars still aligned (despite the exposures being dithered) as if #2179 never happened.

Screenshot 2024-02-22 162201

[🐱](https://jira.stsci.edu/browse/JDAT-4224)

### PR DESCRIPTION
**Pull request by pllim**
### Description

I corrected the test first. Will push commit(s) out later when I can figure out how to fix it properly.

Fixes #2724

### Change log entry

- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [x] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Did the CI pass? If not, are the failures related?
- [x] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone. Bugfix milestone also needs an accompanying backport label.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)? [🐱](https://jira.stsci.edu/browse/JDAT-4224) [👻](https://jira.stsci.edu/browse/JDAT-4342)

### CODE REVIEW

**Comment by pllim** [comment]
Is this going in circles?

https://github.com/spacetelescope/jdaviz/blob/deda388472169d11604db304a369869e2e13d00e/jdaviz/configs/imviz/plugins/orientation/orientation.py#L410-L413

triggers

https://github.com/spacetelescope/jdaviz/blob/deda388472169d11604db304a369869e2e13d00e/jdaviz/app.py#L546

that is being listened back in

https://github.com/spacetelescope/jdaviz/blob/deda388472169d11604db304a369869e2e13d00e/jdaviz/configs/imviz/plugins/orientation/orientation.py#L122-L123

that calls this

https://github.com/spacetelescope/jdaviz/blob/deda388472169d11604db304a369869e2e13d00e/jdaviz/configs/imviz/plugins/orientation/orientation.py#L437

that would trigger this again?

https://github.com/spacetelescope/jdaviz/blob/deda388472169d11604db304a369869e2e13d00e/jdaviz/configs/imviz/plugins/orientation/orientation.py#L410-L413

**Comment by rosteen** [comment]
> Is this going in circles?

I think it could potentially get called one extra time, but the second time `self.orientation.selected` wouldn't actually change so it wouldn't trigger another loop.

**Comment by pllim** [comment]
Theoretically Orientation is the only plugin that is allowed to change the ref data, right? There is no reason for it to listen to its own message?

**Comment by rosteen** [comment]
You can also change the ref data through the viewer data menu, so the Orientation plugin needs to see those events to stay in synch.

**Comment by pllim** [comment]
Right... 😒  Thanks for the reminder!

**Comment by rosteen** [comment]
Drats, I thought there was a good chance that this also fixed the bug I was trying to fix on Friday, but it didn't. I did confirm that it fixes the intended use case though. I'll take a look at the actual code before approving.

**Comment by pllim** [comment]
> the actual code

Sorry for the messy diff.

File: jdaviz/configs/imviz/plugins/orientation/orientation.py [hunk 1]
```diff
@@ -515,3 +521,222 @@ def _update_layer_label_default(self, event={}):
...
+    if link_type == 'wcs':
+        at_least_one_data_have_wcs = len([
+            hasattr(d, 'coords') and isinstance(d.coords, BaseHighLevelWCS)
+            for d in app.data_collection
+        ]) > 0
+        if not at_least_one_data_have_wcs:
+            if wcs_fallback_scheme is None:
+                if error_on_fail:
+                    raise ValueError("link_type can only be 'wcs' when wcs_fallback_scheme "
+                                     "is 'None' if all data have valid WCS.")
```
Comment (rosteen) [inline jdaviz/configs/imviz/plugins/orientation/orientation.py:586]: It looks like we check that at least one loaded image has a WCS, but this error message is saying all images must have WCS. The error message should match the actual check.

**Review by rosteen — COMMENTED** [review]
Looks good to me, I tested a couple more edge cases. I left one comment about an error message but other than that the code looks fine.

**Comment by pllim** [comment]
Code suggestion accepted. Thanks!

**Review by rosteen — APPROVED** [review]
LGTM now.

**Review by cshanahan1 — COMMENTED** [review]
I tested this out in ImvizDitheredExample and it works well for me. 

There was one instance when I was messing around with it that when you did some combination of linking, adding and removing viewers, etc, that after that it because very laggy (particularly with subset creation, they were taking 5+ seconds to appear after drawing), but I can't seem to recreate that same scenario.

Would it be possible to add additional test coverage? A lot of the lines are not covered

**Comment by pllim** [comment]
OK looks like the lines are some safeguards that didn't get triggered by test suite. I can look into that as part of this PR but I should probably rebase first. 🤞

File: jdaviz/configs/default/plugins/markers/tests/test_markers_plugin.py [hunk 2]
```diff
@@ -130,6 +130,7 @@ def test_markers_cubeviz(cubeviz_helper, spectrum1d_cube):
 
     # appears as option in export plugin and exports successfully
     assert "Markers:table" in exp.table.choices
+    exp.filename = str(tmp_path / "cubeviz_export.ecsv")
```
Comment (pllim) [inline jdaviz/configs/default/plugins/markers/tests/test_markers_plugin.py:133]: xref #2755

**Comment by pllim** [comment]
Change log failure is unrelated but something I need to fix anyway: https://github.com/scientific-python/action-check-changelogfile/issues/13

**Comment by pllim** [comment]
Clare verbally approved the PR at tag up this morning, so I am just going to merge when CI is green.

**Comment by cshanahan1** [comment]
haha yes i retroactively approve :)

**Pull request merged**